### PR TITLE
Add canal validation to route finding

### PR DESCRIPTION
**Pull request by ron-murhammer**
## Overview
With the release of RSOC v3, the lack of canal validation during route finding is very apparent. This has been true for all maps but most have only a handful of canals where RSOC has hundreds. Route finding should do some basic canal checks to ensure the canal is controlled and at least 1 of the selected units can move through.

## Functional Changes
- Route finding now does some basic canal validation to ensure at least 1 of the selected units can pass through the canal in order to consider using it on the route
- Canal attachment validation moved to map parsing rather than route finding...
- Improved performance of checking if a canal is on a route by avoiding looping through every territory in the map

## Manual Testing Performed
- Tested mostly on RSOC but checked a few standard maps like revised and global to ensure no issues
- Performance was initially awful but after a couple iterations should now seem to be the same as before to regular users
- Tested editing a map XML to have only 1 of the pair of canal attachments specified to ensure the error is caught during map parsing

### Before
image
image

### After
image
image

### CODE REVIEW

File: game-core/src/main/java/games/strategy/engine/data/GameMap.java [hunk 1]
```diff
@@ -223,6 +224,17 @@ public Territory getTerritory(final String s) {
...
+      final Collection<Unit> units, final PlayerID player) {
+    final Set<Territory> neighbors = getNeighbors(territory, neighborFilter);
+    if (player != null) {
+      return neighbors.parallelStream()
+          .filter(t -> MoveValidator.canAnyUnitsPassCanal(territory, t, units, player, getData()))
+          .collect(Collectors.toSet());
+    }
+    return neighbors;
+  }
+
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/engine/data/GameMap.java]: Parallel stream did seem to make some performance improvement here. This most likely should be used in a lot of other places especially in the AI but gotta start somewhere.

File: game-core/src/main/java/games/strategy/engine/data/RouteFinder.java [hunk 2]
```diff
@@ -12,18 +13,32 @@
...
+    player = null;
+  }
+
+  RouteFinder(final GameMap map, final Predicate<Territory> condition, final Collection<Unit> units,
+      final PlayerID player) {
+    this.map = map;
+    this.condition = condition;
+    previous = new HashMap<>();
+    this.units = units;
+    this.player = player;
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/engine/data/RouteFinder.java]: I don't particularly like how this class works but decided not to make major refactoring changes here.

File: game-core/src/main/java/games/strategy/triplea/attachments/CanalAttachment.java [hunk 3]
```diff
@@ -31,24 +32,29 @@ public CanalAttachment(final String name, final Attachable attachable, final Gam
     super(name, attachable, gameData);
   }
 
-  public static Set<Territory> getAllCanalSeaZones(final String canalName, final GameData data) {
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/attachments/CanalAttachment.java]: This method was one of the big performance problems as it checked every territory for canal attachments. So it was replaced with a method to pass in the route to check for the given canal.

File: game-core/src/main/java/games/strategy/triplea/attachments/CanalAttachment.java [hunk 4]
```diff
@@ -31,24 +32,29 @@ public CanalAttachment(final String name, final Attachable attachable, final Gam
...
         }
       }
+      if (previousTerritoryHasCanal && currentTerritoryHasCanal) {
+        return true;
+      }
+      previousTerritoryHasCanal = currentTerritoryHasCanal;
     }
-    if (territories.size() != 2) {
-      throw new IllegalStateException(
-          "Wrong number of sea zones for canal (exactly 2 sea zones may have the same canalName):" + territories);
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/attachments/CanalAttachment.java]: This validation was moved down to the parsing validate method as it should be checked during map parsing not every time a route is calculated.

File: game-core/src/main/java/games/strategy/triplea/delegate/MoveValidator.java [hunk 5]
```diff
@@ -268,6 +268,28 @@ private static String validateCanal(final Route route, final Collection<Unit> un
     return checkLandTransports(data, player, potentialLandTransports, unitsToLandTransport).isEmpty() ? null : result;
   }
 
+  /**
+   * Simplified version of {@link #validateCanal(Route, Collection, PlayerID, GameData) validateCanal} used for route
+   * finding to check neighboring territories and avoid validating every unit. Performance of this method is critical.
+   */
+  public static boolean canAnyUnitsPassCanal(final Territory start, final Territory end, final Collection<Unit> units,
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/delegate/MoveValidator.java]: This is the core of the implementation which is a simplified canal validation method that checks if the canal is controlled and at least 1 unit can pass through. This makes it much faster than a full validation.

File: game-core/src/main/java/games/strategy/triplea/delegate/MoveValidator.java [hunk 6]
```diff
@@ -1251,37 +1273,24 @@ private static MoveValidationResult validateParatroops(final boolean nonCombat,
     return result;
   }
 
-  /*
-   * Checks if route is either null or includes both canal territories so needs to be checked.
-   */
-  private static boolean isCanalOnRoute(final CanalAttachment canalAttachment, final Route route, final GameData data) {
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/delegate/MoveValidator.java]: This was combined into the CanalAttachment method.

File: game-core/src/main/java/games/strategy/triplea/delegate/MoveValidator.java [hunk 7]
```diff
@@ -1443,20 +1452,16 @@ public static Route getBestRoute(final Territory start, final Territory end, fin
     final boolean isNeutralsImpassable =
         isNeutralsImpassable(data) || (hasAir && !Properties.getNeutralFlyoverAllowed(data));
     final Predicate<Territory> noNeutral = Matches.territoryIsNeutralButNotWater().negate();
-    final Predicate<Territory> noAa = Matches.territoryHasEnemyAaForFlyOver(player, data).negate();
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/delegate/MoveValidator.java]: The changes here are mostly just small refactorings to improve readability and then calling the new `getRouteIgnoreEndValidatingCanals` method.

File: game-core/src/main/java/games/strategy/engine/data/RouteFinder.java [hunk 8]
```diff
@@ -12,18 +13,32 @@
...
+  private final Collection<Unit> units;
+  private final PlayerID player;
 
   RouteFinder(final GameMap map, final Predicate<Territory> condition) {
     this.map = map;
     this.condition = condition;
     previous = new HashMap<>();
+    units = new HashSet<>();
+    player = null;
+  }
```
Comment (ssoloff) [inline game-core/src/main/java/games/strategy/engine/data/RouteFinder.java]: With the addition of the second constructor, can the body of this constructor simply be the following?

```java
this(map, condition, new HashSet<>(), null);
```

File: game-core/src/main/java/games/strategy/triplea/attachments/CanalAttachment.java [hunk 9]
```diff
@@ -154,6 +160,22 @@ public void validate(final GameData data) throws GameParseException {
     if (m_landTerritories == null || m_landTerritories.size() == 0) {
       throw new GameParseException("Canal named " + m_canalName + " must have landTerritories set!" + thisErrorMsg());
     }
+    final Set<Territory> territories = new HashSet<>();
+    for (final Territory t : data.getMap()) {
+      final Set<CanalAttachment> canalAttachments = get(t);
+      if (canalAttachments.isEmpty()) {
+        continue;
+      }
```
Comment (ssoloff) [inline game-core/src/main/java/games/strategy/triplea/attachments/CanalAttachment.java]: It's late, so I may not be reading this properly, but is this `if/continue` necessary?  If `canalAttachments` is empty, won't the for-each loop below simply do nothing?  That is, is the following equivalent?

```java
    final Set<Territory> territories = new HashSet<>();
    for (final Territory t : data.getMap()) {
      for (final CanalAttachment canalAttachment : get(t)) {
        if (canalAttachment.getCanalName().equals(m_canalName)) {
          territories.add(t);
        }
      }
    }
```

**EDIT:** I see now that this is legacy code that was just copy/pasted, so if you don't feel comfortable changing it, don't worry about it.

**Comment by Cernelius** [comment]
Cool, but, marginal and purely visual matter, I wonder if this is letting the default path being the one across the canal if I'm Ctrl it either both before and after the canal, or only after the canal, if it is a direct movement (my unit is currently right in front of the canal).

For example, in your case of the Panama canal, I agree that choosing the impossible path for not enough movement is preferable over choosing the impossible path for the impassable canal, but I wanted just to point out, in case, that if I would be hitting Ctrl on 20 Sea Zone, then it should select the direct path, and telling me why I cannot pass that canal. What I mean is that, if I'm in 19 Sea Zone and hover the pointer on 20 Sea Zone, the selected path should be the 19-18...21-20, but, then, if I Ctrl click on 20 Sea Zone, the selected path (I believe) should be the direct 1 movement to 20 Sea Zone, with the info why I cannot pass the canal, just like you would get it now.

As another case, if I would be in 18 Sea Zone and would use Ctrl click on first 19 Sea Zone then 20 Sea Zone, the impossible path selected should be the one through the impassable canal, not the path sending me back to 18 Sea Zone and around all the way to 20 Sea Zone.

Basically, what I'm saying is that in all cases there are two zones that are connected with each other, if my unit is in one of those and I Ctrl in the other or I used Ctrl in sequence for both, then I think it should select that path anyways (and telling me why I can't, in case).

As another example, if I'm in 11 Sea Zone and I Ctrl click first on 19 Sea Zone and, then, on 20 Sea Zone, that should be the only selected path, as I've used Ctrl in the full sequence of all direct connections, not the 11-19-18...21-20 path, and not even if I would have the 6 movement points required to make it. On the other hand if, still from 11 Sea Zone, instead, I Ctrl clicked on 19 Sea Zone but then I hover on 20 Sea Zone, without (yet) Ctrl clicking on it, the selected path should be the 11-19-18...21-20 one, also if not enough movement.

No idea if it already works this way. Just wanted to point out, in case.

**Comment by Cernelius** [comment]
Practically, I'm suggesting a special exception that only when, either:
1. using Ctrl in sequence on both ends of a direct connection, or
2. using Ctrl on a zone that is directed connected to the zone where the selected unit is,

then the selected path is always and forcefully the single direct movement between these zones, never trying to get around it indirectly, not even if possible (so to allow specifically selecting an, even impossible, path, by using Ctrl in the full sequence of movements).

On the other hand, if using Ctrl in any other cases, the selected path would still be one of the shortest ones amongst the possible ones (not blocked by canals, territory effects...), with only the condition of also passing through the specified zones. For example, if I'm in 11 Sea Zone and only Ctrl click on 20 Sea Zone, the path would be the 11-18...21-20 one, but, instead, if I Ctrl click on 19 Sea Zone and, then, 20 Sea Zone, the path would be the 11-19-20 one, not the 11-19-18...21-20 one.

Just a suggestion to have a way to force the engine to select exactly the path you want to (also if impossible, that has the minor value of getting the info why you cannot go that path, in the bottom bar). Of course, it would need also to be documented that when using Ctrl for a direct connection (either your unit being on the other end of it or having just used Ctrl on the other end of it), then only the direct path is possibly defined.

**Comment by ron-murhammer** [comment]
@Cernelius It already works that way currently and extends to these changes as well. So for the revised example above, if I CTRL click on sz20 even though it is initially showing going around South America, it will then show sz19 > sz20 directly.

**Comment by Cernelius** [comment]
>@Cernelius It already works that way currently and extends to these changes as well. So for the revised example above, if I CTRL click on sz20 even though it is initially showing going around South America, it will then show sz19 > sz20 directly.

In this case, I suggest documenting such a behaviour in:

Help:Movement/Selection Help

What you currently see there is only:
>CTRL-Left click on a territory to select the territory as a way point (this will force the units to move through this territory on their way to the destination).

Based on the above, if the destroyer is in 19 Sea Zone and I Ctrl click in 20 Sea Zone, I should expect it still showing all the way around 19-18...21-20, as nothing in the description hints otherwise.
If a continuous CTRL path also has the effect of strictly defining such path, and not only assuring the selected territory to be part of it, this special behaviour should be clarified to the users.

File: game-core/src/main/java/games/strategy/engine/data/RouteFinder.java (hunk 8 above)
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/engine/data/RouteFinder.java]: Done.

File: game-core/src/main/java/games/strategy/triplea/attachments/CanalAttachment.java (hunk 9 above)
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/attachments/CanalAttachment.java]: Yeah, didn't touch this code but I'm fine making that small improvement. Done.

**Comment by ron-murhammer** [comment]
Alright. I added a small clarification to CTRL moving to indicate it always takes the shortest path to the way point as well (existing behavior).

**Comment by Cernelius** [comment]
Sorry for asking again, since I cannot follow your development or looking at the code, but, reading the updated info help, does that mean that the general behaviour changes to always selecting one of the shortest paths in any case Ctrl is used at any point?

For example, if  I'm in 11 Sea Zone and I only Ctrl click on 20 Sea Zone (but not setting the entire exact path by previously clicking on 19 Sea Zone), will it, then, select the 11-19-20 path, instead of 11-18...21-20?

In this case, if the above is correct, maybe it would be clearer as:

"CTRL-Left click on a territory to select the territory as a way point and prioritize a shortest path to destination, even if not valid (this will force the units to move through this territory on their way to the destination and doing so via one of the shortest possible paths, even if all are impassable)."

Also, I'm starting to seriously consider that maybe these two behaviours should be distinct, but this would be off topic.

**Pull request merged**